system/src/system_network_manager_api.cpp

```diff
@@ -401,7 +401,7 @@ void manage_network_connection() {
     /* FIXME: refactor */
     if (SPARK_WLAN_RESET || SPARK_WLAN_SLEEP) {
         auto wasSleeping = SPARK_WLAN_SLEEP;
-        cloud_disconnect();
+        cloud_disconnect(CLOUD_DISCONNECT_GRACEFULLY, CLOUD_DISCONNECT_REASON_NETWORK_DISCONNECT);
```

Comment:
I feel we should distinguish between user specified network disconnects, unanticipated network disconnects and the network watchdog disconnects. This will help us know whether the disconnect is intentional or not.

Reply:
:+1: I will need to add a new field to the Goodbye message and update its handler in Device Service accordingly.

Reply:
I added a new field that contains network disconnection reason to Goodbye message and updated the system cloud API accordingly. It should be noted that, currently, the only network disconnection reason that can be reported via Goodbye message is `NETWORK_DISCONNECT_REASON_USER`, since, normally, the system itself doesn't trigger a network disconnection, and when it does (e.g. to recover from recurring cloud connection errors), it's not connected to the cloud anyway.